src/MongoDB/Query.c

```diff
@@ -50,64 +50,244 @@ PHONGO_API zend_class_entry *php_phongo_query_ce;
+	if (zarr && php_array_existsc(zarr, key)) { \
+		if (!php_phongo_query_opts_append_document(intern->opts, (opt), (zarr), (key) TSRMLS_CC)) { \
+			return false; \
+		} \
+	}
 
-			if (Z_TYPE_P(zread_concern) != IS_OBJECT || !instanceof_function(Z_OBJCE_P(zread_concern), php_phongo_readconcern_ce TSRMLS_CC)) {
-				phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Expected \"readConcern\" option to be %s, %s given", ZSTR_VAL(php_phongo_readconcern_ce->name), zend_get_type_by_const(Z_TYPE_P(zread_concern)));
-				bson_clear(&intern->query);
-				return;
+	/* Note: handling of integer options will depend on SIZEOF_ZEND_LONG and we
+	 * are not converting strings to 64-bit integers for 32-bit platforms. */
+
+#define PHONGO_QUERY_OPT_INT64(opt, zarr, key) \
+	if (zarr && php_array_existsc(zarr, key)) { \
+		if (!BSON_APPEND_INT64(intern->opts, (opt), php_array_fetchc_long(zarr, key))) { \
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Error appending \"%s\" option", (opt)); \
+			return false; \
+		} \
+	}
+
+#define PHONGO_QUERY_OPT_STRING(opt, zarr, key) \
+	if (zarr && php_array_existsc(zarr, key)) { \
+		if (!php_phongo_query_opts_append_string(intern->opts, (opt), (zarr), (key) TSRMLS_CC)) { \
+			return false; \
+		} \
+	}
+
+	PHONGO_QUERY_OPT_BOOL("allowPartialResults", options, "allowPartialResults");
+	PHONGO_QUERY_OPT_BOOL("awaitData", options, "awaitData");
+	PHONGO_QUERY_OPT_INT64("batchSize", options, "batchSize");
+	PHONGO_QUERY_OPT_DOCUMENT("collation", options, "collation");
+	PHONGO_QUERY_OPT_STRING("comment", options, "comment") else PHONGO_QUERY_OPT_STRING("comment", modifiers, "$comment");
+	PHONGO_QUERY_OPT_BOOL("exhaust", options, "exhaust");
+	PHONGO_QUERY_OPT_DOCUMENT("max", options, "max") else PHONGO_QUERY_OPT_DOCUMENT("max", modifiers, "$max");
+	PHONGO_QUERY_OPT_INT64("maxScan", options, "maxScan") else PHONGO_QUERY_OPT_INT64("maxScan", modifiers, "$maxScan");
+	PHONGO_QUERY_OPT_INT64("maxTimeMS", options, "maxTimeMS") else PHONGO_QUERY_OPT_INT64("maxTimeMS", modifiers, "$maxTimeMS");
+	PHONGO_QUERY_OPT_DOCUMENT("min", options, "min") else PHONGO_QUERY_OPT_DOCUMENT("min", modifiers, "$min");
+	PHONGO_QUERY_OPT_BOOL("noCursorTimeout", options, "noCursorTimeout");
+	PHONGO_QUERY_OPT_BOOL("oplogReplay", options, "oplogReplay");
+	PHONGO_QUERY_OPT_DOCUMENT("projection", options, "projection");
+	PHONGO_QUERY_OPT_BOOL("returnKey", options, "returnKey") else PHONGO_QUERY_OPT_BOOL("returnKey", modifiers, "$returnKey");
+	PHONGO_QUERY_OPT_BOOL("showRecordId", options, "showRecordId") else PHONGO_QUERY_OPT_BOOL("showRecordId", modifiers, "$showDiskLoc");
+	PHONGO_QUERY_OPT_INT64("skip", options, "skip");
+	PHONGO_QUERY_OPT_DOCUMENT("sort", options, "sort") else PHONGO_QUERY_OPT_DOCUMENT("sort", modifiers, "$orderby");
+	PHONGO_QUERY_OPT_BOOL("snapshot", options, "snapshot") else PHONGO_QUERY_OPT_BOOL("snapshot", modifiers, "$snapshot");
+	PHONGO_QUERY_OPT_BOOL("tailable", options, "tailable");
```

Comment:
Because this is a hard coded list, it would be necessary to add code if new options are going to be supported by libmongoc/the server. Do we want to add a fall back mechanism for this?